### PR TITLE
rec: Disable outgoing v4 when QLA has no v4 addresses

### PR DESCRIPTION
**Pull request by pieterlexis**
### Short description
The documentation actually mentions that not configuring a certain address family in `query-local-address` means no queries are sent out from that address family. This was only true for IPv4 though. This PR ensures this is true for v6 only installations as well.

### Checklist
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

File: pdns/pdns_recursor.cc [hunk 1]
```diff
@@ -4102,6 +4102,15 @@ static int serviceMain(int argc, char*argv[])
     exit(99);
   }
 
+  if(pdns::isQueryLocalAddressFamilyEnabled(AF_INET)) {
+    SyncRes::s_doIPv4=true;
+    g_log<<Logger::Warning<<"Enabling IPv4 transport for outgoing queries"<<endl;
+  }
+  else {
+    g_log<<Logger::Warning<<"NOT using IPv6 for outgoing queries - add an IPv4 address (like '0.0.0.0') to query-local-address to enable"<<endl;
```
Comment (phonedph1) [inline pdns/pdns_recursor.cc:4110]: s/IPv6/IPv4/ ?

File: pdns/recursordist/test-syncres_cc10.cc [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+  sr->setAsyncCallback([target, &v4Hit, &v6Hit, &queries](const ComboAddress& ip, const DNSName& domain, int type, bool doTCP, bool sendRDQuery, int EDNS0Level, struct timeval* now, boost::optional<Netmask>& srcmask, boost::optional<const ResolveContext&> context, LWResult* res, bool* chained) {
+    queries++;
+    if (isRootServer(ip)) {
+      setLWResult(res, 0, false, false, true);
+
+      if (domain == DNSName("powerdns.com.")) {
+        addRecordToLW(res, domain, QType::NS, "ns1.powerdns.com.", DNSResourceRecord::AUTHORITY, 172800);
+        addRecordToLW(res, "ns1.powerdns.com.", QType::AAAA, "2001:DB8:1::53", DNSResourceRecord::ADDITIONAL, 3600);
+        addRecordToLW(res, "ns1.powerdns.com.", QType::A, "192.0.2.1", DNSResourceRecord::ADDITIONAL, 3600);
+      }
```
Comment (rgacogne) [inline pdns/recursordist/test-syncres_cc10.cc:27]: Don't we have a 50% chance of passing this test even if the "no v4" switch is broken? Perhaps we could check that the root server address used is v4 as well?

File: pdns/recursordist/test-syncres_cc10.cc [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+  std::unique_ptr<SyncRes> sr;
+  initSR(sr);
+  SyncRes::s_doIPv4 = false;
+  SyncRes::s_doIPv6 = true;
+  primeHints();
+  int queries = 0;
+
+  const DNSName target("powerdns.com.");
+  sr->setAsyncCallback([target, &queries](const ComboAddress& ip, const DNSName& domain, int type, bool doTCP, bool sendRDQuery, int EDNS0Level, struct timeval* now, boost::optional<Netmask>& srcmask, boost::optional<const ResolveContext&> context, LWResult* res, bool* chained) {
+    cout<<ip.toString()<<endl;
```
Comment (rgacogne) [inline pdns/recursordist/test-syncres_cc10.cc:109]: I'm pretty sure the `cout` lines are not wanted here :-)

**Comment by pieterlexis** [comment]
The failing test is because that test uses `query-local-address=::1` and the auths don't listen on ipv6.

**Comment by omoerbeek** [comment]
Rebased and fixed a little bug that prevented error exit in unthreaded mode after failure to load hints.

**Comment by omoerbeek** [comment]
@rgacogne could you check the new exit code from serviceMain? There are more cases to be converted later, but we first need to make sure the approach is right.

File: pdns/pdns_recursor.cc [hunk 4]
```diff
@@ -179,6 +179,7 @@ struct RecThreadInfo
   bool isListener{false};
   /* process queries */
   bool isWorker{false};
+  int exitCode{0};
```
Comment (rgacogne) [inline pdns/pdns_recursor.cc:182]: It would be nicer to move that above the booleans and below the `uint64_t` :-)

**Pull request merged**